mulighetsrommet-api/src/main/kotlin/no/nav/mulighetsrommet/api/avtaler/AvtaleValidator.kt

```diff
@@ -0,0 +1,138 @@
+import arrow.core.raise.either
+import arrow.core.raise.ensure
+import arrow.core.right
+import no.nav.mulighetsrommet.api.domain.dbo.AvtaleDbo
+import no.nav.mulighetsrommet.api.repositories.AvtaleRepository
+import no.nav.mulighetsrommet.api.repositories.TiltaksgjennomforingRepository
+import no.nav.mulighetsrommet.api.repositories.TiltakstypeRepository
+import no.nav.mulighetsrommet.api.routes.v1.responses.ValidationError
+import no.nav.mulighetsrommet.domain.Tiltakskoder.isTiltakMedAvtalerFraMulighetsrommet
+import no.nav.mulighetsrommet.domain.constants.ArenaMigrering
+import no.nav.mulighetsrommet.domain.dto.AvtaleAdminDto
+import no.nav.mulighetsrommet.domain.dto.Avtalestatus
+import no.nav.mulighetsrommet.domain.dto.TiltakstypeDto
+
+class AvtaleValidator(
+    private val tiltakstyper: TiltakstypeRepository,
+    private val avtaler: AvtaleRepository,
+    private val tiltaksgjennomforinger: TiltaksgjennomforingRepository,
+) {
+    fun validate(dbo: AvtaleDbo): Either<List<ValidationError>, AvtaleDbo> = either {
+        val tiltakstype = tiltakstyper.get(dbo.tiltakstypeId)
+            ?: raise(ValidationError("tiltakstype", "Tiltakstypen finnes ikke").nel())
+
+        val errors = buildList {
+            if (!dbo.startDato.isBefore(dbo.sluttDato)) {
+                add(ValidationError("startDato", "Startdato må være før sluttdato"))
+            }
+
+            if (dbo.navEnheter.isEmpty()) {
+                add(ValidationError("navEnheter", "Minst ett NAV-kontor må være valgt"))
+            }
+
+            if (dbo.leverandorUnderenheter.isEmpty()) {
+                add(ValidationError("leverandorUnderenheter", "Minst én underenhet til leverandøren må være valgt"))
+            }
+
+            avtaler.get(dbo.id)?.also { avtale ->
+                ensure(avtale.avtalestatus in listOf(Avtalestatus.Planlagt, Avtalestatus.Aktiv)) {
+                    plus(
+                        ValidationError(
+                            "navn",
+                            "Kan bare gjøre endringer når avtalen har status Planlagt eller Aktiv",
+                        ),
+                    )
+                }
+
+                ensure(dbo.opphav == avtale.opphav) {
+                    plus(ValidationError("opphav", "Avtalens opphav kan ikke endres"))
+                }
+
+                val (numGjennomforinger) = tiltaksgjennomforinger.getAll(avtaleId = dbo.id)
+                if (numGjennomforinger > 0) {
+                    if (dbo.tiltakstypeId != avtale.tiltakstype.id) {
+                        add(
+                            ValidationError(
+                                "tiltakstypeId",
+                                "Kan ikke endre tiltakstype fordi det finnes gjennomføringer for avtalen",
+                            ),
+                        )
+                    }
+                    // TODO validere at start- og slutt-dato passer med gjennomføringer... men ta høyde for dette ikke gjelder AFT/VTA?
```

Comment:
Skal denne gjøres?

Reply:
Var litt usikker, siden det skurrer litt for VTA/AFT. Jeg fjerner TODO'en, men implementerer det som en ny oppgave i stedet.